src/node/node_class.hpp

```diff
@@ -297,111 +1330,195 @@ template<typename ClassType>
+	if (instanceProxy.IsUndefined()) {
+		instanceProxy = info.This().As<Napi::Object>();
+	}
+
+	try {
+		F(env, instanceProxy, args, result);
+		return result.ToValue();
+	}
+	catch (const Napi::Error& e) {
+		throw;
+	}
+	catch (const node::Exception& e) {
+		Napi::Error error = Napi::Error::New(info.Env(), e.what());
+		copyObject(env, e.m_value, error);
+		throw error;
+	}
+	catch (const std::exception& e) {
+		throw Napi::Error::New(info.Env(), e.what());
+	}
 }
 
-
 template<node::PropertyType::GetterType F>
-void wrap(v8::Local<v8::String> property, const Nan::PropertyCallbackInfo<v8::Value>& info) {
-    v8::Isolate* isolate = info.GetIsolate();
-    node::ReturnValue return_value(info.GetReturnValue());
-    try {
-        F(isolate, info.This(), return_value);
-    }
-    catch (std::exception &e) {
-        Nan::ThrowError(node::Exception::value(isolate, e));
-    }
+Napi::Value wrap(const Napi::CallbackInfo& info) {
+	Napi::Env env = info.Env();
+	node::ReturnValue result(env);
+	try {
+		F(env, info.This().As<Napi::Object>(), result);
+		return result.ToValue();
+	}
+	catch (const Napi::Error& e) {
+		throw;
+	}
+	catch (const node::Exception& e) {
+		Napi::Error error = Napi::Error::New(info.Env(), e.what());
+		copyObject(env, e.m_value, error);
+		throw error;
+	}
+	catch (const std::exception& e) {
+		throw Napi::Error::New(info.Env(), e.what());
+	}
 }
 
 template<node::PropertyType::SetterType F>
-void wrap(v8::Local<v8::String> property, v8::Local<v8::Value> value, const Nan::PropertyCallbackInfo<void>& info) {
-    v8::Isolate* isolate = info.GetIsolate();
-    try {
-        F(isolate, info.This(), value);
-    }
-    catch (std::exception &e) {
-        Nan::ThrowError(node::Exception::value(isolate, e));
-    }
+void wrap(const Napi::CallbackInfo& info, const Napi::Value& value) {
+	Napi::Env env = info.Env();
+
+	try {
+		F(env, info.This().As<Napi::Object>(), value);
+	}
+	catch (const Napi::Error& e) {
+		throw;
+	}
+	catch (const node::Exception& e) {
+		Napi::Error error = Napi::Error::New(info.Env(), e.what());
+		copyObject(env, e.m_value, error);
+		throw error;
+	}
+	catch (const std::exception& e) {
+		throw Napi::Error::New(info.Env(), e.what());
+	}
 }
 
 template<node::IndexPropertyType::GetterType F>
-void wrap(uint32_t index, const Nan::PropertyCallbackInfo<v8::Value>& info) {
-    v8::Isolate* isolate = info.GetIsolate();
-    node::ReturnValue return_value(info.GetReturnValue());
-    try {
-        F(isolate, info.This(), index, return_value);
-    }
-    catch (std::out_of_range &) {
-        // Out-of-bounds index getters should just return undefined in JS.
-        return_value.set_undefined();
-    }
-    catch (std::exception &e) {
-        Nan::ThrowError(node::Exception::value(isolate, e));
-    }
+Napi::Value wrap(const Napi::CallbackInfo& info, const Napi::Object& instance, uint32_t index) {
+	Napi::Env env = info.Env();
+	node::ReturnValue result(env);
+
+	try {
+		try {
```

Comment:
Why do we need nested `try` blocks?

Reply:
1. code readability. A bit more obvious the parent try handles the error handling while the nested try handles the logic error
2. inside the nested catch block we have more code which might throw an exception which will be caught by the parent try..catch block.